### ISSUE TITLE
Unable to perform Level 2: Basic Authorization

### ISSUE THREAD

**Issue by thomaskoeck** [authzforce/fiware#8]
Hi,

I need some help please: I want to perform Level 2 basic authorization. So I created a new role in my application on keyrock (idm) and created two new permissions: One that has the HTTP verb GET and /test1 as resource and one that has POST as HTTP verb and /test2 as resource, but I did not yet assign these permissions to my newly created role. I created these two resources just for testing purposes. All they do is that they send back a text message that tell me if I could access these resources. My newly created role I assigned to one of my registered users.

Using Chrome's Advanced REST Client I sent a GET and POST request for these resources to the pep-proxy. For both requests I got a response of 401 Unauthorized, which is fine since I did not include a X-Auth-Token in these requests yet. Then I performed the authentication with the oauth2 example-client and got back my token. I copied the token into the header field and sent the same requests to the pep-proxy again. In both cases I got back a 200 OK message and the dedicated success messages that I created. But actually this should not be the case. Instead I should get back a 401 Unauthorized messages since the role of the user, which I am logged in as, does not have the permissions to access these resources.

Why can I still access these resources? It seems to me the only thing that is checked is if the token is valid or not. As soon as the token is valid, the user can access whatever he wants. Did I do something wrong?

I run everything as docker containers. Here is some log output for the GET request:

pep | 2016-11-25 12:44:05.300 - INFO: IDM-Client - Token in cache, checking timestamp...
pep | 2016-11-25 12:44:05.300 - INFO: IDM-Client - Token in cache expired
pep | 2016-11-25 12:44:05.300 - INFO: IDM-Client - Checking token with IDM...
keyrock | 2016-11-25 12:44:05.331 34 INFO eventlet.wsgi.server [-] 172.18.0.7 - - [25/Nov/2016 12:44:05] "GET /v3/access-tokens/nalLDoB334Z3BItu0ytcoUJOmOC3m2 HTTP/1.1" 200 394 0.026148
pep | 2016-11-25 12:44:05.332 - INFO: AZF-Client - Checking auth with AZF...
pep | 2016-11-25 12:44:05.332 - INFO: AZF-Client - Checking authorization to roles [ '5fedd57e74c94a9b993db26b145c1035' ] to do GET on test1 and app eb5fc491be0d4edd946cc6ce20a096b3
pep | 2016-11-25 12:44:05.332 - INFO: AZF-Client - Checking auth with AZF...
pep | 2016-11-25 12:44:05.345 - INFO: Root - Access-token OK. Redirecting to app...

I hope someone can help me with this.

Best regards,
Thomas

**Comment by cdanger** [comment]
Hello,
which versions of KeyRock, PEP Proxy and AuthzForce are you using?

**Comment by thomaskoeck** [comment]
I am using Authzforce release-5.4.1, KeyRock v5.3.0 and latest version of PEP Proxy.

**Comment by cdanger** [comment]
Please upgrade KeyRock to v5.4.0. Some important AuthzForce integration issues have been fixed in this release. (Also I assume you are using PEP Proxy v5.4.) Let me know if you are able to reproduce the issue with this version. Besides, make sure that when you define a new role (and associated permissions), you click the Save button at the bottom, and as a result, you get a status message below like "Modified roles and permissions", which confirms your role-based access control policies has been pushed to AuthzForce.

**Comment by thomaskoeck** [comment]
I can't even log in to KeyRock v5.4.0 using username "idm" and password "idm". I get a KeyError.

Traceback:
File "/horizon/.venv/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  87.                 response = middleware_method(request)
File "/horizon/openstack_dashboard/fiware_middleware/middleware.py" in process_request
  155.         if (fiware_api.keystone.user_is_password_expired(request)):
File "/horizon/openstack_dashboard/fiware_api/keystone.py" in user_is_password_expired
  276.     return (datetime.datetime.today() - password_changed_date).days > settings.FIWARE_DEFAULT_DURATION['user_password']

Exception Type: KeyError at /
Exception Value: 'user_password' 

In KeyRock v5.3.0 I can login without any problems.

**Comment by cdanger** [comment]
You should raise this issue to KeyRock development team (which I am **not** a member of) on their Github.
In the meantime, if you want to stick with KeyRock v5.3.0, which I do not recommend on the long term, then you have to apply these two fixes on your instance by yourself:
- [72](https://github.com/ging/fiware-idm/issues/72): you can check my [pull request](https://github.com/ging/horizon/pull/2/files) to know how to fix it; basically fixing a typo in file `openstack_dashboard/templates/access_control/policy.xacml`. 
- [71](https://github.com/ging/fiware-idm/issues/71): see my [pull request](https://github.com/ging/horizon/pull/3/files) as well to get it fixed. 

Enable debug all in the file `log_config.json`:

``` javascript
{
  "appenders": [
    { 
      "type": "console",
      "layout": {
        "type": "pattern",
        "pattern": "%d  - %p: %c - %m",
        "replaceConsole": true
      } 
    }
  ],
  "levels": {
    "Server": "DEBUG",
    "Root": "DEBUG",
    "HTTP-Client": "DEBUG",
    "AZF-Client": "DEBUG",
    "IDM-Client": "DEBUG",
    "Test": "DEBUG"
  }
}
```

Then restart the IdM container after fixing the files. Then try to reproduce the issue. Make sure that when you define a new role (and associated permissions), you click the Save button at the bottom, and as a result, you get a status message below like "Modified roles and permissions", which confirms your role-based access control policies has been pushed to AuthzForce.
If this goes wrong, dump the debug logs here.

**Comment by thomaskoeck** [comment]
Sorry for my late reply. With the latest version of KeyRock it works now.

Thanks for your help.

**Comment by cdanger** [comment]
Excellent!

**Comment by sanchit31** [comment]
hi @thomaskoeck,

I'm facing the same issue.
The user is able to access all the pages of the application, even if the user has no permission to access it.

Keyrock version: 7.4.0
Wilma PEP Proxy Server 7.4.0
AuthZForce : 8.0.1

Is there way to check whether roles and permissions that have been set in IdM, are updated in AuthZForce too?

**Comment by cdanger** [comment]
You can perform this kind of HTTP request on AuthzForce Server's REST API to find out ([more info](https://authzforce-ce-fiware.readthedocs.io/en/latest/UserAndProgrammersGuide.html#domain-management-api)): 
```
GET /domains?externalId=my-app-123
HTTP/1.1
Accept: application/xml; charset=UTF-8
```

Replace `my-app-123` with the application ID in Keyrock.
The response should look like this:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<resources
  xmlns="http://authzforce.github.io/rest-api-model/xmlns/authz/5"
  xmlns:atom="http://www.w3.org/2005/Atom">
  <atom:link rel="item" href="XXX" title="XXX"/>
</resources>
```

If there is a (atom) `link` element, it means a domain was created in AuthzForce successfully, and the domain ID is `XXX` as given by the `href` attribute. If there is none, something wrong happened in the IdM, so the AuthzForce domain was not created (or you did not click the Submit/Save button on the permissions page).
Once you have the AuthzForce domain ID, e.g. `XXX`, you can check the list of policies corresponding to the roles and permissions on AuthzForce as follows:

```
GET /domains/XXX/pap/policies
HTTP/1.1
Accept: application/xml; charset=UTF-8
```

**Comment by sanchit31** [comment]
hi @cdanger 

Thank you for your help.

I get the following response when I do this 
`GET /domains/XXX/pap/policies
HTTP/1.1
Accept: application/xml; charset=UTF-8`

`<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ns4:resources xmlns="http://authzforce.github.io/core/xmlns/pdp/6.0" xmlns:ns2="http://authzforce.github.io/pap-dao-flat-file/xmlns/properties/3.6" xmlns:ns3="http://www.w3.org/2005/Atom" xmlns:ns4="http://authzforce.github.io/rest-api-model/xmlns/authz/5" xmlns:ns5="urn:oasis:names:tc:xacml:3.0:core:schema:wd-17">
    <ns3:link rel="item" href="root"/>
</ns4:resources>`

I can also see the following policy created in data/domains directory in authzforce
`<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PolicySet xmlns="urn:oasis:names:tc:xacml:3.0:core:schema:wd-17" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	PolicySetId="root" Version="0.1.0"
	PolicyCombiningAlgId="urn:oasis:names:tc:xacml:3.0:policy-combining-algorithm:deny-unless-permit">
	<Target />
	<Policy RuleCombiningAlgId="urn:oasis:names:tc:xacml:3.0:rule-combining-algorithm:deny-unless-permit" PolicyId="permit-all" Version="0.1.0">
		<Target />
		<Rule Effect="Permit" RuleId="permit-all" />
	</Policy>
</PolicySet>`

I'm setting the rules and permission for the user to access /resource1 of the application but the user is able to access all the resources of the application.

Also I get this ### **warning : authzforce error** in IdM while mapping roles to permissions in IdM.

**Comment by cdanger** [comment]
The IdM fails to push your rules/permissions to AuthzForce. Try raise the IdM log level to DEBUG in order to get more details about this "authzforce error". As far as I can see, the issue is in the IdM. Therefore, if the IdM debug logs are not good enough, you should ask the IdM dev team for support.

**Comment by letavia** [comment]
Hi @sanchit31 did you manage to find the solution? I got the same problem and have tried pretty much all suggestions there are.

**Comment by sanchit31** [comment]
hi @letavia 
Unfortunately I haven't been able to find its solution. 
Do you see any error or warning in the IdM logs?

**Comment by letavia** [comment]
@sanchit31 That's the problem, I didn't see any errors in the IdM log. It is just the IdM seems to not push the rules to AZF even when I created the domain manually in AZF to match the app ID in the IdM management console.  
I tried another setup with IdM as PDP but with no luck either. The authentication works fine, but afterwards PEP just allows access to any resources, even when the authorization decision shows "deny" in IdM, seemingly PEP does not bother to check the authorization. I dont know if there's a default rule which allows all access that needs to be overridden/removed before the authorization works.
